File: app/src/main/java/com/lostincontext/data/rules/HeadphoneFenceVM.java
package com.lostincontext.data.rules;


import android.support.annotation.NonNull;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.google.android.gms.awareness.fence.AwarenessFence;

import java.util.List;

public class HeadphoneFenceVM implements FenceVM {

    public enum State {
        PLUGGED_IN,
        PLUGGED_OUT
    }

    private final @NonNull State state;

    @JsonCreator
    public HeadphoneFenceVM(@JsonProperty("state") @NonNull State state) {
        this.state = state;
    }

    @Override
    public AwarenessFence build(FenceBuilder builder) {
        return builder.headphone(this);
    }

    @Override public String describe(FenceDescriptor descriptor) {
        return descriptor.headphone(this);
    }

    @Override public void giveIcon(FenceIconGiver iconGiver, List<Integer> icons) {
        iconGiver.headphone(this, icons);
    }

    @NonNull public State getState() {
        return state;
    }

}

File: app/src/main/java/com/lostincontext/data/rules/HeadphoneFenceVM.kt
package com.lostincontext.data.rules


import com.google.android.gms.awareness.fence.AwarenessFence

data class HeadphoneFenceVM(val state: State) : FenceVM {

    enum class State {
        PLUGGED_IN,
        PLUGGED_OUT
    }


    override fun build(builder: FenceBuilder): AwarenessFence = builder.headphone(this)

    override fun describe(descriptor: FenceDescriptor): String = descriptor.headphone(this)

    override fun giveIcon(iconGiver: FenceIconGiver,
                          icons: MutableList<Int>): Unit = iconGiver.headphone(this, icons)
}
